### PR TITLE
Dynamic dashboard #24

### PR DESCRIPTION
**Pull request by joelmackenz**
### What this PR does (required):
- Adds dynamic rendering of profiles into ProfileListings.tsx, including search by city and search by available days functions.
- Adds routes and controllers to handle this searching on the backend
- Adds ProfileListings > daySelect to replace the MUI calendar previously implemented (so we can easily only choose days of the week)

### Screenshots / Videos (required):

https://user-images.githubusercontent.com/66138947/122123964-7f2f2b00-cde3-11eb-8a42-8544ab4c29b5.mov

### Any information needed to test this feature (required):
- Fast way: Import "ProfileListings.tsx" into client > App.js and replace Routes with ProfileListings

### Any issues with the current functionality (optional):
- In ProfileListings.js, I set a temporary list holding info from the API request, and set its type to "any". Of course, I would rather set these types to "Profile" or "User", but when I do so, an error is thrown warning that the result may be undefined, which is incompatible. I have checks which ensure that the result will not be undefined, and attempted changing the interface to allow for undefined as an option, but the error persists. The "any" type works, it just bothers me. (I left notes on ProfileListings.js showing where these occur.)

### Notes
- In server > controllers > profile.js, returned users are limited to 100. This is to ensure the FE is not overloaded. Of course, eventually, we would need functionality to display more than 100 if we had so many user accounts.
- Search for days allows for users available in any of the given days. This could eventually also allow for searching for specific day availability (ie: only show me users who are available on all Mon, Tues, *and* Wed)

### CODE REVIEW

File: client/src/context/interface/Profile.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import { User } from './User';
+
+export interface Profile {
+  _id: string;
+  isDogSitter: { type: Boolean; require: true; default: false };
+  isAvailable: { type: Boolean; require: true; default: false };
```
Comment (moffatethan) [inline client/src/context/interface/Profile.ts]: This is a bit confusing. Why not just allow `isDogSitter: boolean;`. Require and defaults are all handled by Mongo but in terms of the data we pass down, it just have to be the type set on the model. So `isDogSitter: boolean;` still gives the same behaviour and more understandable.

File: client/src/context/interface/Profile.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import { User } from './User';
+
+export interface Profile {
```
Comment (moffatethan) [inline client/src/context/interface/Profile.ts]: There is an interfaces folder in the `src/` directory, there is no need to create a new directory for interfaces.

Comment (joelmackenz) [reply client/src/context/interface/Profile.ts]: Oops! Thank you.

File: client/src/pages/ProfileListings/ProfileListings.tsx [hunk 3]
```diff
@@ -52,31 +52,109 @@ export default function ProfileListings(): JSX.Element {
...
+            searchResults.push(user.profile);
+          }
+        });
+        setDisplayedProfiles(searchResults);
+      }
     }
   };
 
   const profileCardGrid = (
     <Grid container className={classes.profilesContainer} xs={9}>
```
Comment (moffatethan) [inline client/src/pages/ProfileListings/ProfileListings.tsx]: Grid containers do not accept the `xs` property or any grid sizing property. Those only are for columns.

File: client/src/pages/ProfileListings/ProfileListings.tsx [hunk 4]
```diff
@@ -52,31 +52,109 @@ export default function ProfileListings(): JSX.Element {
...
-      {displayedUsers.map((user) => (
-        <ProfileCard user={user} key={user._id} />
+      {displayedProfiles.map((profile, index) => (
+        <ProfileCard profile={profile} key={`${profile._id}-${index}`} />
```
Comment (moffatethan) [inline client/src/pages/ProfileListings/ProfileListings.tsx]: The key here probably doesn't need the index, the MongoID won't be duplicated in the back-end.

File: server/controllers/profile.js [hunk 5]
```diff
@@ -3,72 +3,187 @@ const asyncHandler = require("express-async-handler");
...
+                    } else {
+                        res.json({ success: true, msg: "Message added" });
+                    }
+                }
+            );
+        });
+    } catch (err) {
+        console.log(err);
+        res.send(err);
+    }
```
Comment (moffatethan) [inline server/controllers/profile.js]: Express-async-handler will deal with catching thrown errors meaning the try/catch is unnecessary here.

File: server/routes/profile.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
+const express = require("express");
+const router = express.Router();
+
+const auth = require("../middleware/auth");
+
+// Connect to profile controller
+const profileController = require("../controllers/profile");
+
+/* POST route to create profile within a user object - POST PROFILE operation. */
```
Comment (moffatethan) [inline server/routes/profile.js]: The code is pretty clear what it does, these kind of comments are unnecessary.

**Review by moffatethan — CHANGES_REQUESTED** [review]
Hello! Left you some comments, otherwise. Looking good!

File: server/routes/profile.js (hunk 6 above)
Comment (joelmackenz) [reply server/routes/profile.js]: Good point. Thanks.

**Review by moffatethan — APPROVED** [review]
LGTM! Thank you!

**Please fix merge conflicts before merging!**

**Pull request merged**